python/ray/data/datasource/iceberg_datasource.py

```diff
@@ -0,0 +1,187 @@
+        self._scan_kwargs = scan_kwargs if scan_kwargs is not None else {}
+
+        self.table_identifier = table_identifier
+
+        self._catalog_type = catalog_type
+        self._catalog_kwargs = catalog_kwargs if catalog_kwargs is not None else {}
+
+        self._row_filter = row_filter if row_filter is not None else AlwaysTrue()
+        self._selected_fields = selected_fields
+        self._snapshot_id = snapshot_id
+
+    def _get_catalog(self) -> "Catalog":
+        from pyiceberg import catalog
+
+        if self._catalog_type == "glue":
+            # Glue catalogs have no name, and no conf options are needed as long as AWS
+            # credentials are set up
+            return catalog.load_glue(name="", conf=self._catalog_kwargs)
+        else:
+            raise NotImplementedError(
+                f"Catalog type {self._catalog_type} not implemented"
+            )
+
+    def _get_table_and_data_scan(self) -> Tuple["Table", "DataScan"]:
+        catalog = self._get_catalog()
+        table = catalog.load_table(self.table_identifier)
+
+        data_scan = table.scan(
+            row_filter=self._row_filter,
+            selected_fields=self._selected_fields,
+            snapshot_id=self._snapshot_id,
+            **self._scan_kwargs,
+        )
+
+        return table, data_scan
+
+    def estimate_inmemory_data_size(self) -> Optional[int]:
+        # Get the plan files in this query
+        _, data_scan = self._get_table_and_data_scan()
+        return sum(task.length for task in data_scan.plan_files())
+
+    @staticmethod
+    def _distribute_tasks_into_equal_chunks(
+        plan_files: Iterable["FileScanTask"], n_chunks: int
+    ) -> List[List["FileScanTask"]]:
+        """
+        Implement a greedy knapsack algorithm to distribute the files in the scan
+        across tasks, based on their file size, as evenly as possible
+        """
+        chunks = [list() for _ in range(n_chunks)]
+        chunk_sizes = {chunk: 0 for chunk in range(n_chunks)}
+
+        # From largest to smallest, add the plan files ot the smallest chunk one at a
+        # time
+        for plan_file in sorted(plan_files, key=lambda f: f.length, reverse=True):
+            smallest_chunk = min(chunk_sizes, key=chunk_sizes.get)
+            chunks[smallest_chunk].append(plan_file)
+            # Update the size of the chunk after adding the file to it
+            chunk_sizes[smallest_chunk] += plan_file.length
+
+        return chunks
+
+    def get_read_tasks(self, parallelism: int) -> List[ReadTask]:
+        from pyiceberg.io import pyarrow as pyi_pa_io
+
+        def _get_read_task(
+            tasks: Iterable["FileScanTask"],
+            table_identifier: str,
+            schema: "Schema",
+            catalog_type: str = "glue",
+        ) -> Iterable[block.Block]:
+            # Closure so we can pass this callable as an argument to ReadTask
+
+            # Both the catalog and tbl attributes cannot be pickled, which means they
```

Comment:
I think we should be able to do this:

- https://github.com/apache/iceberg-python/issues/513
- https://github.com/apache/iceberg-python/issues/514

Reply:
@Fokko, do we want to be able to pickle a table and a catalog? There are definitely reasons to want to ship a table to another process, but we've found that this can also enable people to do problematic things accidentally, like planning scans in parallel. Not being able to pickle ensures that planning happens once and tasks (which should be picklable) are sent instead.

Reply:
That's a good question. I think `Table` makes sense, but I can see that the `Catalog` is less obvious:

- **Table**: You can already serialize the FileIO today. I think being able to serialize the Table makes sense since this is more-or-less just serializing the TableMetadata. We can reconstruct it on the instance itself.
- **Catalog**: This might also make sense if we want to refresh credentials and such, but looking at the code here, we could also avoid pickling the catalog.